torch/csrc/jit/export.cpp

```diff
@@ -16,222 +17,177 @@ namespace torch { namespace jit {
 
 namespace {
 
-namespace onnx = ::torch::onnx;
+namespace onnx_torch = ::torch::onnx;
+namespace onnx = ::ONNX_NAMESPACE;
 
-std::string value_name(Value* n) {
-  return n->uniqueName();
+std::string getNodeStackTraceString(const Node* n) {
+  std::stringstream ss;
+  if (n->getSourceLocation()) {
+    n->getSourceLocation()->highlight(ss);
+  } else {
+    ss << "<unknown location>";
+  }
+  return ss.str();
 }
 
-struct ExportContext {
-  size_t num_blocks = 0;
-  onnx::OperatorExportTypes operator_export_type;
-};
-
-void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g,
-                 const std::vector<at::Tensor> & initializers,
-                 ExportContext *ctx, RawDataExportMap* raw_data_export_map=nullptr);
+void validateGraph(const std::shared_ptr<Graph>& graph, onnx_torch::OperatorExportTypes operator_export_type) {
+  for (auto node : graph->nodes()) {
+      // Macro'ed so we get a marginally better line number on failed export
+#define FAIL_EXPORT(name) \
+      throw std::runtime_error(std::string("ONNX export failed: ") + name + "\n\nGraph we tried to export:\n" + graph->toString());
+    IR_IF(node, PythonOp)
+      auto py_node = static_cast<torch::jit::PythonOp*>(value);
+      FAIL_EXPORT(
+          "Couldn't export Python operator " + py_node->name() +
+          "\n\nDefined at:\n" + getNodeStackTraceString(node))
+    IR_ELSE()
+      // Special error messages for certain types of operators
+      if (node->kind() == aten::expand) {
+        FAIL_EXPORT(
+            "Could not export a broadcasted operation; ONNX likely does not support this form of broadcasting.\n\nBroadcast occurred at:\n" +
+            getNodeStackTraceString(node));
+      }
+      if (node->kind() == prim::PackPadded || node->kind() == prim::PadPacked) {
+        FAIL_EXPORT(
+            "Cannot export individual pack_padded_sequence or pad_packed_sequence; these operations must occur in pairs.\n\nUsage of this operation occurred at:\n" +
+            getNodeStackTraceString(node));
+      }
+      bool is_aten_fallback = operator_export_type == onnx_torch::OperatorExportTypes::ONNX_ATEN_FALLBACK;
+      if (!node->kind().is_onnx() && !is_aten_fallback && node->kind() != prim::Undefined) {
+        FAIL_EXPORT(
+            "Couldn't export operator " + node->kind().toDisplayString() + "\n\nDefined at:\n" +
+            getNodeStackTraceString(node));
+      }
+    IR_END()
+#undef FAIL_EXPORT
+  }
+}
 
-void encodeBlock(onnx::GraphProto * p_g, Block *b,
-                const std::vector<at::Tensor> & initializers,
-                ExportContext *ctx, RawDataExportMap* raw_data_export_map);
+class JitEncoder {
```

Comment:
maybe call it EncoderBase? Jit is kinds of a confusing name (even though it's all over the place)